### PR TITLE
Changes applied in Author and Expression to use Guard.Against in Cons…

### PR DESCRIPTION
**Pull request by ShadyNagy**
Changes applied in Author and Expression to use Guard.Against in Constructor and unit test added for Author and Expression.

### CODE REVIEW

**Comment by ardalis** [comment]
Good stuff but I agree with @antmdvs on both points:
- guard clauses first before any other logic
- turn on NRTs and then re-assess what guards and tests you need

https://devblogs.microsoft.com/dotnet/try-out-nullable-reference-types/

**Comment by ShadyNagy** [comment]
> 
> 
> Guard clauses are typically stacked at the top of methods so processing fails fast if preconditions aren't met.
> 
did you mean by "Nullable Reference Types" change
```csharp 
[CanBeNull] public string Title { get; private set; }
```
to 
```csharp 
public string Title { get; private set; }
```
then I can remove all Guard clauses so we can use outside the entity.

> My personal preference would be for Nullable Reference Types here. I think you'll find that you'll be able to delete 7 of your unit tests because they'll simply fail to compile with NRT enabled.

**Comment by antmdvs** [comment]
You have to enable the NRT language feature using one of the methods in the linked article. You can enable it for the whole project by adding `<Nullable>enable</Nullable>` in the .csproj file. As soon as you do that, you can lean on the compiler (it will show squiggles) to delete the tests where you're passing `null` to a non-nullable parameter. You can subsequently delete unnecessary guard clauses.

**Comment by ShadyNagy** [comment]
Already done.
> 
> 
> @ShadyNagy You have to enable the NRT language feature using one of the methods in the linked article. You can enable it for the whole project by adding `<Nullable>enable</Nullable>` in the .csproj file. As soon as you do that, you can lean on the compiler (it will show squiggles) to delete the tests where you're passing `null` to a non-nullable parameter. You can subsequently delete unnecessary guard clauses.

**Comment by antmdvs** [comment]
@ShadyNagy It looks like you have to enable NRT in the test project as well in order to see the warnings:

image

File: src/RegExLib.Core/Entities/Author.cs [hunk 1]
```diff
@@ -1,6 +1,7 @@
 ﻿using System.Collections.Generic;
 using System.Collections.ObjectModel;
 using Ardalis.GuardClauses;
+using JetBrains.Annotations;
```
Comment (antmdvs) [inline src/RegExLib.Core/Entities/Author.cs:4]: Was this added by accident?

Comment (ShadyNagy) [reply src/RegExLib.Core/Entities/Author.cs]: JetBrains.Annotations;

> 
> 
> Was this added by accident?

**Comment by antmdvs** [comment]
The `?` makes the reference types nullable again. You can remove those from the property declarations.

**Comment by ShadyNagy** [comment]
I added <Nullable>enable</Nullable> in unit test without any effect 
to remove this warning I need to add "!"
like that 
```csharp
Assert.Throws<ArgumentNullException>(() => new Author(null!, "userAdmin", "Admin Admin"));
```
applying to all not.

> @ShadyNagy It looks like you have to enable NRT in the test project as well in order to see the warnings:

**Comment by antmdvs** [comment]
> I added `<Nullable>enable</Nullable>` in unit test and it does not make any effect
> to remove this warning I need to add "!"

The idea is that you can **delete** those tests now because with NRT, you're now getting compile-time guarantees that those arguments won't be null, at least if the calling code has also opted into NRT.

@ardalis I think we should elevate the NRT warnings to errors if possible. Would you agree?

File: src/RegExLib.Core/Entities/Author.cs [hunk 2]
```diff
@@ -29,9 +31,8 @@ public void AddExpression(Expression expression)
             _expressions.Add(expression);
         }
 
-        public override string ToString()
+        public override string? ToString()
         {
-            Guard.Against.NullOrWhiteSpace(FullName, nameof(FullName));
             return FullName;
         }
```
Comment (antmdvs) [inline src/RegExLib.Core/Entities/Author.cs:37]: I would remove the `?` on the return type.

Is `FullName` truly nullable? If so, I'd default the return value to an empty string or other fallback value as a courtesy to the calling code, so it doesn't have to perform a null check:

```csharp
public override string ToString() => Fullname ?? "";
```

If not, this should do:

```csharp
public override string ToString() => Fullname;
```

Comment (ShadyNagy) [reply src/RegExLib.Core/Entities/Author.cs]: @antmdvs Done

**Comment by antmdvs** [comment]
I realize the NRT stuff is a little confusing. By adding `?` everywhere, you've effectively gone back to the default reference type behavior, which would require guard clauses everywhere. It might help to think about this feature as enabling _NON_-Nullable Reference Types. I think it's called "Nullable Reference Types" because it requires a `?` (just like a nullable value type) to indicate to the compiler that the type can be null. 

From here, I'd look at all the entities' properties and remove the `?` from the ones that are required for the entity and therefore should never be `null`. I'm assuming all the ctor params correspond to required properties at this point. For any such parameters, you can also remove any `Guard.Against.Null()` statements. Finally, any unit test that is expecting `ArgumentNullException` can be deleted because the compiler will warn you by design if you try to pass a null to something that should not accept a null. 

Hope that helps

**Comment by ShadyNagy** [comment]
@antmdvs I need your suggestion on that.
That is cleared now.
but what about this 
```csharp
        private static void Validate(string userId, string username)
        {
            Guard.Against.NullOrWhiteSpace(userId, nameof(userId));
            Guard.Against.NullOrWhiteSpace(username, nameof(username));
        }
```
I think we needed to check username and userId inside the Entity or that also needs to remove?
and if yes need to remove 
what we will do in this
```csharp
Username = username?.Replace("@", "");
```

> 
> 
> I realize the NRT stuff is a little confusing. By adding `?` everywhere, you've effectively gone back to the default reference type behavior, which would require guard clauses everywhere. It might help to think about this feature as enabling _NON_-Nullable Reference Types. I think it's called "Nullable Reference Types" because it requires a `?` (just like value types) to indicate to the compiler that the type can be null.
> 
> From here, I'd look at all the entities' properties and remove the `?` from the ones that are required for the entity and therefore should never be `null`. I'm assuming all the ctor params correspond to required properties at this point. For any such parameters, you can also remove any `Guard.Against.Null()` statements. Finally, any unit test that is expecting `ArgumentNullException` can be deleted because the compiler will warn you by design if you try to pass a null to something that should not accept a null.
> 
> Hope that helps

**Comment by ShadyNagy** [comment]
I think as you said we should elevate the NRT warnings to errors.

**Comment by ShadyNagy** [comment]
@antmdvs I tried to apply a rule to make NRT warnings to errors 

but I have 2 tests have problem in RegExLib.UnitTests
```csharp
        [Fact]
        public async Task ThrowsExceptionGivenNullEventArgument()
        {
            Exception ex = await Assert.ThrowsAsync<ArgumentNullException>(() => _handler.Handle(null, CancellationToken.None));
        }
```
```csharp
        public Task<object> Send(object request, CancellationToken cancellationToken = default)
        {
            return Task.FromResult<object>(default);
        }
```

**Comment by ShadyNagy** [comment]
And if yes then we need to apply all of that or only (CS8600, CS8601)?

```
CS8073 | The result of the expression is always the same since a value of this type is never equal to ‘null’
CS8597 | Thrown value may be null.
CS8600 | Converting null literal or possible null value to non-nullable type.
CS8601 | Possible null reference assignment.
CS8602 | Dereference of a possibly null reference.
CS8603 | Possible null reference return.
CS8604 | Possible null reference argument.
CS8605 | Unboxing a possibly null value.
CS8606 | Possible null reference assignment to iteration variable
CS8607 | A possible null value may not be passed to a target marked with the [DisallowNull] attribute
CS8608 | Nullability of reference types in type doesn’t match overridden member.
CS8609 | Nullability of reference types in return type doesn’t match overridden member.
CS8610 | Nullability of reference types in type of parameter doesn’t match overridden member.
CS8611 | Nullability of reference types in type of parameter doesn’t match partial method declaration.
CS8612 | Nullability of reference types in type doesn’t match implicitly implemented member.
CS8613 | Nullability of reference types in return type doesn’t match implicitly implemented member.
CS8614 | Nullability of reference types in type of parameter doesn’t match implicitly implemented member.
CS8615 | Nullability of reference types in type doesn’t match implemented member.
CS8616 | Nullability of reference types in return type doesn’t match implemented member.
CS8617 | Nullability of reference types in type of parameter doesn’t match implemented member.
CS8618 | Non-nullable field is uninitialized. Consider declaring as nullable.
CS8619 | Nullability of reference types in value doesn’t match target type.
CS8620 | Argument cannot be used for parameter due to differences in the nullability of reference types.
CS8621 | Nullability of reference types in return type doesn’t match the target delegate.
CS8622 | Nullability of reference types in type of parameter doesn’t match the target delegate.
CS8624 | Argument cannot be used as an output for parameter due to differences in the nullability of reference types.
CS8625 | Cannot convert null literal to non-nullable reference type.
CS8626 | The ‘as’ operator may produce a null value for a type parameter.
CS8629 | Nullable value type may be null.
CS8631 | The type cannot be used as type parameter in the generic type or  method. Nullability of type argument doesn’t match constraint type.
CS8632 | The annotation for nullable reference types should only be used in code within a ‘#nullable’ annotations context.
CS8633 | Nullability in constraints for type parameter doesn’t match the  constraints for type parameter in implicitly implemented interface  method’.
CS8634 | The type cannot be used as type parameter in the generic type or  method. Nullability of type argument doesn’t match ‘class’ constraint.
CS8638 | Conditional access may produce a null value for a type parameter.
CS8643 | Nullability of reference types in explicit interface specifier doesn’t match interface implemented by the type.
CS8644 | Type does not implement interface member. Nullability of reference  types in interface implemented by the base type doesn’t match.
CS8645 | Interface is already listed in the interface list with different nullability of reference types.
CS8653 | A default expression introduces a null value for a type parameter.
CS8654 | A null literal introduces a null value for a type parameter.
CS8655 | The switch expression does not handle some null inputs.
CS8667 | Partial method declarations have inconsistent nullability in constraints for type parameter
CS8714 | The type cannot be used as type parameter in the generic type or  method. Nullability of type argument doesn’t match ‘notnull’ constraint.
```

**Comment by ardalis** [comment]
Yes I'd turn on warnings as errors for all projects.

**Comment by ardalis** [comment]
if there are a few areas that are blowing up but are outside of what you're working on, you can disable nullable just around those areas for now and add an issue for someone (perhaps me) to fix them.

```
#nullable disable
// nullable won't apply here
# nullable enable
```

**Comment by ShadyNagy** [comment]
RegExLib.Core - nullable enabled
RegExLib.Infrastructure - nullable enabled
RegExLib.SharedKernel - nullable disable -> version is not C# 8
RegExLib.Web - nullable enabled
RegExLib.FunctionalTests - nullable enabled
RegExLib.IntegrationTests - nullable enabled
RegExLib.UnitTests - nullable enabled

**Comment by antmdvs** [comment]
I'm pretty we could achieve the same with an .editorconfig file too, but unfortunatlely:

> EditorConfig files can only be used to enable rules and set their severity in **Visual Studio 2019 version 16.3 and later**.

(https://docs.microsoft.com/en-us/visualstudio/code-quality/analyzers-faq?view=vs-2019#editorconfig-versus-rule-sets)

so we should probably stick with the .csproj/MSBuild method since other contributors may be on a lower version of VS.

**Comment by ShadyNagy** [comment]
I think the ruleset will be better as it is only one file applied to all projects and also maybe we need to add something else in the future in the ruleset.

> 
> 
> I'm not sure we need the ruleset. I think you can put this in the .csproj files instead:
> 
> ```cs
> <WarningsAsErrors>nullable</WarningsAsErrors>`
> ```
> 
> [dotnet/roslyn#41605 (comment)](https://github.com/dotnet/roslyn/issues/41605#issuecomment-585258085)

**Comment by ShadyNagy** [comment]
@ardalis  All issues added for enabled nullable.

**Comment by ardalis** [comment]
Sounds good, thanks!

**Pull request merged**